scipy/stats/_binned_statistic.py

```diff
@@ -446,21 +564,31 @@ def binned_statistic_dd(sample, values, statistic='mean',
                 null = np.nan
             np.seterr(**old)
         result.fill(null)
-        for i in np.unique(xy):
-            result[i] = statistic(values[xy == i])
+        for i in np.unique(binnumbers):
+            for vv in range(Vdim):
+                result[vv, i] = statistic(values[vv, binnumbers == i])
 
     # Shape into a proper matrix
-    result = result.reshape(np.sort(nbin))
+    result = result.reshape(np.append(Vdim, np.sort(nbin)))
+
     for i in np.arange(nbin.size):
         j = ni.argsort()[i]
-        result = result.swapaxes(i, j)
+        # Accomodate the extra `Vdim` dimension-zero with `+1`
+        result = result.swapaxes(i+1, j+1)
         ni[i], ni[j] = ni[j], ni[i]
 
     # Remove outliers (indices 0 and -1 for each dimension).
-    core = D * [slice(1, -1)]
+    core = [slice(None)] + Ndim * [slice(1, -1)]
     result = result[core]
 
-    if (result.shape != nbin - 2).any():
+    # Unravel binnumbers into an ndarray, each row the bins for each dimension
+    if(expand_binnumbers and Ndim > 1):
+        binnumbers = np.asarray(np.unravel_index(binnumbers, nbin))
+
+    if np.any(result.shape[1:] != nbin - 2):
         raise RuntimeError('Internal Shape Error')
 
-    return BinnedStatisticddResult(result, edges, xy)
+    # Remove excess dimension-zero if ``Vdim == 1``
+    result = result.squeeze()
```

Comment:
This will produce an unexpected result if the user inputs a length-1 list of arrays. I think we need to save the input shape above, and use that here to choose the right output.